joint_trajectory_controller/src/joint_trajectory_controller.cpp

```diff
@@ -85,31 +90,154 @@ JointTrajectoryController::update()
+        rt_active_goal_->setFeedback(feedback);
+
+        // check abort
+        if (abort) {
+          // RCLCPP_WARN(lifecycle_node_->get_logger(), "Aborted due to state tolerance violation");
+          // auto result = std::make_shared<FollowJTrajAction::Result>();
+          // result->set__error_code(FollowJTrajAction::Result::PATH_TOLERANCE_VIOLATED);
+          // rt_active_goal_->setAborted(result);
+          // rt_active_goal_.reset();
+        }
+
+        // check goal tolerance
+        if (!before_last_point) {
+          if (!outside_goal_tolerance) {
+            auto res = std::make_shared<FollowJTrajAction::Result>();
+            res->set__error_code(FollowJTrajAction::Result::SUCCESSFUL);
+            rt_active_goal_->setSucceeded(res);
+            rt_active_goal_.reset();
+
+            RCLCPP_INFO(lifecycle_node_->get_logger(), "Goal reached, success!");
+          } else if (default_tolerances_.goal_time_tolerance != 0.0) {
+            // if we exceed goal_time_toleralance set it to aborted
+            rclcpp::Time traj_start = (*traj_point_active_ptr_)->get_trajectory_start_time();
+            rclcpp::Time traj_end = traj_start + start_segment_itr->time_from_start;
+
+            double difference = lifecycle_node_->now().seconds() - traj_end.seconds();
+            if (difference > default_tolerances_.goal_time_tolerance) {
+              // auto result = std::make_shared<FollowJTrajAction::Result>();
+              // result->set__error_code(FollowJTrajAction::Result::GOAL_TOLERANCE_VIOLATED);
+              // rt_active_goal_->setAborted(result);
+              // rt_active_goal_.reset();
+              /*RCLCPP_WARN(
+                lifecycle_node_->get_logger(),
+                "Aborted due goal_time_tolerance exceeding by %f seconds (given: %f)",
+                difference, default_tolerances_.goal_time_tolerance);*/
+            }
+          }
+        }
+      }
     }
 
-    prev_traj_point_ptr_ = traj_point_ptr;
     set_op_mode(hardware_interface::OperationMode::ACTIVE);
   }
+  else
+  {
+    state_desired = traj_msg_home_ptr_->points[0];
 
-  publish_state();
+    for (auto index = 0ul; index < joint_num; ++index) {
+        compute_error_for_joint(state_error, index, state_current, state_desired);
+    }
+  }
+  
+  double period = lifecycle_node_->get_parameter("period").as_double();
+  //convert to an effort command buffer via ClosedLoopHardwareInterfaceAdapter::updateCommand
+  for (auto index = 0ul; index < joint_num; ++index) {
+    
+    double f = 0.0;
+    if (index < state_error.positions.size() && index < state_error.velocities.size() && index < state_desired.velocities.size())
+    {
+      f = pids_[index]->computeCommand(state_error.positions[index], state_error.velocities[index], rclcpp::Duration::from_seconds(period));
+
+      const double command = (state_desired.velocities[index] * velocity_ff_[index]) + f;
+      registered_joint_cmd_handles_[index]->set_cmd(command);
```

Comment:
It took a significant amount of effort tracing down effort_controllers::JointTrajectoryController. But it boiled down to this code. I use control_toolbox::pid to convert the errors to effort values.